File: thuc_hanh_3/3.2.2.py
# Name: Nguyễn Minh Hoàng
# Student ID: 202418904
# Class: 763965

"""Thực hành 3 - Tráo đổi hai giá trị qua tham chiếu khả biến."""

from dataclasses import dataclass
import sys


sys.stdout.reconfigure(encoding="utf-8")


@dataclass
class ThamChieuSo:
    gia_tri: float


def trao_doi(so_thu_nhat: ThamChieuSo, so_thu_hai: ThamChieuSo) -> None:
    so_thu_nhat.gia_tri, so_thu_hai.gia_tri = so_thu_hai.gia_tri, so_thu_nhat.gia_tri


def main() -> None:
    try:
        so_thu_nhat = ThamChieuSo(float(input("Nhập số thứ nhất: ")))
        so_thu_hai = ThamChieuSo(float(input("Nhập số thứ hai: ")))
    except ValueError:
        print("Dữ liệu nhập vào phải là số.")
        return

    print(f"Trước khi tráo đổi: a = {so_thu_nhat.gia_tri:g}, b = {so_thu_hai.gia_tri:g}")
    trao_doi(so_thu_nhat, so_thu_hai)
    print(f"Sau khi tráo đổi:   a = {so_thu_nhat.gia_tri:g}, b = {so_thu_hai.gia_tri:g}")


if __name__ == "__main__":
    main()

File: thuc_hanh_3/3.2.2.c
// Name: Nguyễn Minh Hoàng
// Student ID: 202418904
// Class: 763965

/* Thực hành 3 - Phần 2 - Bài 2: tráo đổi hai giá trị bằng con trỏ. */

#include <stdio.h>

void trao_doi(int *so_thu_nhat, int *so_thu_hai) {
    int tam = *so_thu_nhat;
    *so_thu_nhat = *so_thu_hai;
    *so_thu_hai = tam;
}

int main(void) {
    int a;
    int b;

    printf("Nhap so nguyen a: ");
    if (scanf("%d", &a) != 1) {
        printf("Du lieu khong hop le.\n");
        return 1;
    }
    printf("Nhap so nguyen b: ");
    if (scanf("%d", &b) != 1) {
        printf("Du lieu khong hop le.\n");
        return 1;
    }

    printf("Truoc khi trao doi: a = %d, b = %d\n", a, b);
    trao_doi(&a, &b);
    printf("Sau khi trao doi:   a = %d, b = %d\n", a, b);
    return 0;
}
